torch/csrc/distributed/c10d/init.cpp

```diff
@@ -199,18 +199,33 @@ PyObject* c10d_init(PyObject* _unused, PyObject* noargs) {
       .def(
           "get_index",
           &::c10d::GradBucket::getIndex,
-          py::call_guard<py::gil_scoped_release>())
+          py::call_guard<py::gil_scoped_release>(),
+          R"(
+.. warning::
+    Since the buckets are rebuilt after the first iteration, should not rely on the indices at the beginning of training.
+
+Returns:
+    The index of a bucket that stores gradients of a few contiguous layers.
```

Comment:
This can be addressed in a separate PR, but I'm not sure if we should keep `get_Index()` as a public method since gradient bucketing/rebuilding buckets is sort of a DDP internal detail that shouldn't leak to the user.

I see that this is mainly used as key into an error_dict, can we provide a way to hash `GradBucket` instead to use that as a key to the error_dict?

Reply:
Agree that it may unnecessarily expose some internal details.

`bucket_index == 0` is also used to determine if the current bucket is the last one to allreduce in an iteration. This is something hard to be done by a hashcode.